cmd/frontend/internal/app/ui/handlers.go

```diff
@@ -256,6 +257,75 @@ func serveSignIn(w http.ResponseWriter, r *http.Request) error {
 	return renderTemplate(w, "app.html", common)
 }
 
+// redirectTreeOrBlob redirects a blob page to a tree page if the file is actually a directory,
+// or a tree page to a blob page if the directory is actually a file.
+func redirectTreeOrBlob(routeName string, common *Common, r *http.Request, w http.ResponseWriter) (requestHandled bool, err error) {
+	path := mux.Vars(r)["Path"]
+	if path == "/" || path == "" {
+		if routeName != routeRepo {
+			// Redirect to repo route
+			target := "/" + string(common.Repo.Name)
+			if common.Rev != "" {
+				target += common.Rev
+			}
+			http.Redirect(w, r, target, http.StatusTemporaryRedirect)
+			return true, nil
+		}
+		return false, nil
+	}
+	cachedRepo, err := backend.CachedGitRepo(r.Context(), common.Repo)
+	if err != nil {
+		return false, err
+	}
+	stat, err := git.Stat(r.Context(), *cachedRepo, common.CommitID, path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			serveError(w, r, err, http.StatusNotFound)
+			return true, nil
+		}
+		return false, err
+	}
+	expectedDir := (routeName == routeTree)
```

Comment:
I'm not sure, but I think lint will complain (it's currently turned off)
```suggestion
	expectedDir := routeName == routeTree
```